src/fs/metadata_log.hh

```diff
@@ -0,0 +1,252 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+#pragma once
+
+#include "fs/cluster.hh"
+#include "fs/cluster_allocator.hh"
+#include "fs/inode.hh"
+#include "fs/inode_info.hh"
+#include "fs/metadata_disk_entries.hh"
+#include "fs/metadata_to_disk_buffer.hh"
+#include "fs/units.hh"
+#include "fs/unix_metadata.hh"
+#include "fs/value_shared_lock.hh"
+#include "seastar/core/file-types.hh"
+#include "seastar/core/future-util.hh"
+#include "seastar/core/future.hh"
+#include "seastar/core/shared_future.hh"
+#include "seastar/core/shared_ptr.hh"
+#include "seastar/core/sstring.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+#include <chrono>
+#include <cstddef>
+#include <exception>
+#include <type_traits>
+#include <variant>
+
+namespace seastar::fs {
+
+struct invalid_inode_exception : public std::exception {
```

Comment:
It would be better to have a base class for those exceptions so we can catch filesystem-specific exceptions easier. Also we could add error code to those exceptions (see std::filesystem::filesystem_error).

Reply:
What about those error codes, use ones from <cerrno> or invent new ones (in fact, there are not enough specific codes in <cerrno> for our needs).

Reply:
Let's just start by creating a base class for fs exceptions, and leave specific error codes for a later discussion - we'll probably use a mix of standard error codes and some additional errors specific for our implementation.